### ISSUE TITLE
Compiler attributes for improved security

### ISSUE THREAD

**Issue by siddhesh** [ossf/wg-best-practices-os-developers#201]
A number of compiler attributes (e.g. `alloc_size`, `malloc`, `access`, `nonnull`) provide additional information to the compiler to allow it to (1) detect potentially undefined behaviour and (2) optimize code. These often end up in the win-win category where the compiler is able to improve performance as well as security.  I had documented a [couple of them in an article](https://developers.redhat.com/articles/2023/02/06/how-improve-application-security-using-fortifysource3#) in the past in a narrow scope, but there's scope for a more comprehensive document.  There are some caveats to their usage though, so it's likely not a matter of simply listing them.

Would a new document describing these attributes and how OSS developers could use them with best effect be a useful addition to this project?

**Comment by david-a-wheeler** [comment]
The Linux kernel has set up a set of conventions to prevent buffer overwrites in many circumstances, that might be a useful source.

We currently have a project for compiler options, but that's a different thing.

The issue would be that some people would need to dedicate real time for this. If that can be done, I imagine it could be really helpful.

**Comment by siddhesh** [comment]
I don't want to promise (because this is largely going to be free time work for me and it's summer!) but I can take a stab at it at some point. I just wanted to put this issue out here to (1) get a sense for whether this would be interesting for this WG and (2) remind myself of it later whenever I have some free time to tackle it.

**Comment by thomasnyman** [comment]
Fully supporting this! A new document seems to be a sensible approach in my opinion, although there's potentially some overlap with the existing guide too, e.g. [this recent RFC](https://discourse.llvm.org/t/rfc-enforcing-bounds-safety-in-c-fbounds-safety/70854) for Clang proposes a combination of option flag + annotations.

**Comment by david-a-wheeler** [comment]
@siddhesh - could you take a quick stab at a starting document? I think there's general agreement that this is useful, but probably needs to be a separate document. The big difference, as I understand it, is that this involves changes to the source code itself. Feel free to focus on just C (or just C++); I suspect that it may need to be more specific about the language.

I encourage looking at:

* Some of the approaches and discussion of the Linux kernel (current and proposed) 
* GCC C extensions: https://gcc.gnu.org/onlinedocs/gcc/C-Extensions.html

**Comment by siddhesh** [comment]
I can't attend tomorrow's call due to a conflict, but I wanted to share that I've started working on this, here's a preview: https://github.com/siddhesh/wg-best-practices-os-developers/commit/fc1dce2742d0ff049a3e5b73817b96c3ab740acd

I'll send a PR once I do the due diligence, i.e. figure out minimum compiler versions, etc. and write some detail for each of those attributes with notes on their usage.  There may be more attributes to add in there, but that could be done incrementally.

**Comment by SecurityCRob** [comment]
Has this been addressed by the C/C++ Compiler Hardening options guide?    @gkunz @thomasnyman @david-a-wheeler

**Comment by thomasnyman** [comment]
I've picked up work on this in #794 and am planning to at least add detailed descriptions for the currently listed attributes.

We also have #551 in the https://github.com/ossf/wg-best-practices-os-developers/labels/backlog which should be reconsidered at some point.

**Comment by Quuxplusone** [comment]
I came here from https://best.openssf.org/Compiler-Hardening-Guides/Compiler-Annotations-for-C-and-C++.html .
The table on that page (which claims to come from this GitHub repo) claims that "Clang 21.0.0" supports

    int *mymalloc(int) __attribute__((malloc(myfree)));

However, Clang 21 has not yet been released, and Clang trunk (the rising Clang 21) in actuality _doesn't_ support that attribute. [Godbolt](https://godbolt.org/z/Mo6aWYEYM):

```
warning: 'malloc' attribute ignored because Clang does not yet
support this attribute signature [-Wignored-attributes]
    3 | int * __attribute__((malloc(myfree))) mymalloc(int);
      |                      ^
```

Was this an oversight/mistake in #794? Could there be other mistakes in there too? (At least this is the only instance of "Clang 21" on the page!)

**Comment by thomasnyman** [comment]
Thanks @Quuxplusone for reporting this. The Compiler Annotations for C and C++ document is a work-in-progress draft and there can always be mistakes. However, in this case I don't think this is an oversight as the document currently states:

> Clang supports both forms of the malloc attribute but does not yet implement the -Wmismatched-dealloc and -Wmismatched-new-delete warnings. 

As you can see from https://godbolt.org/z/KMsc1nrnz Clang 20.1.0 fails with the error: 

> <source>:6:54: error: 'malloc' attribute takes no arguments

while Clang trunk recognizes the form of the `malloc` attribute with two arguments, but issues the warning you point to. The two-argument form is also documented at https://clang.llvm.org/docs/AttributeReference.html#malloc

While you are of course correct that Clang 21 has not yet been released, it will have been released by the time we intend to stabilize the Compiler Annotations for C and C++ document.

That said, I'm happy to review suggestion on how to clarify the behavior. I'd suggest opening separate issues or PRs with the https://github.com/ossf/wg-best-practices-os-developers/labels/Product%3A%20Compiler%20Annotations%20Guide label to discuss specific rewording.

**Comment by Quuxplusone** [comment]
> Thanks [@Quuxplusone](https://github.com/Quuxplusone) for reporting this. The Compiler Annotations for C and C++ document is a work-in-progress draft and there can always be mistakes. However, in this case I don't think this is an oversight as the document currently states:
> 
> > Clang supports both forms of the malloc attribute but does not yet implement the -Wmismatched-dealloc and -Wmismatched-new-delete warnings.
> 
> As you can see from https://godbolt.org/z/KMsc1nrnz Clang 20.1.0 fails with the error:
> 
> > :6:54: error: 'malloc' attribute takes no arguments
> 
> while Clang trunk recognizes the form of the `malloc` attribute with two arguments, but issues the warning you point to. The two-argument form is also documented at https://clang.llvm.org/docs/AttributeReference.html#malloc

Right, and what I said above was: "Clang trunk (the rising Clang 21) in actuality _doesn't_ support that attribute." Which is correct: Clang doesn't support the two-argument attribute. Look at the warning message from your own Godbolt (emphasis added):

    warning: 'malloc' attribute ignored because Clang does not yet support this attribute signature [-Wignored-attributes]
                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        6 | void *my_malloc(size_t size) __attribute__ ((malloc, malloc (my_free, 1)));
          |                                                      ^

> 
> While you are of course correct that Clang 21 has not yet been released, it will have been released by the time we intend to stabilize the Compiler Annotations for C and C++ document.

Perhaps so; but even Clang 21 will not support that attribute.

> That said, I'm happy to review suggestion on how to clarify the behavior. I'd suggest opening separate issues or PRs with the [ Product: Compiler Annotations Guide](https://github.com/ossf/wg-best-practices-os-developers/labels/Product%3A%20Compiler%20Annotations%20Guide) label to discuss specific rewording.

I'll open a PR.
